modules/videoio/include/opencv2/videoio/utils.private.hpp

```diff
@@ -10,6 +10,83 @@
 
 namespace cv {
 CV_EXPORTS std::string icvExtractPattern(const std::string& filename, unsigned *offset);
+
+class IReadStream
+{
+public:
+    virtual ~IReadStream() {};
+    virtual long long read(char* buffer, long long size) = 0;
+    virtual long long seek(long long offset, int way) = 0;
+    virtual IReadStream* clone() = 0;
+};
+
+class StreambufReadStream : public IReadStream
+{
+public:
+    StreambufReadStream(std::streambuf& _stream) : stream(_stream) {}
+
+    virtual ~StreambufReadStream() {}
+
+    long long read(char* buffer, long long size) override
+    {
+        return stream.sgetn(buffer, size);
+    }
+
+    long long seek(long long offset, int way) override
+    {
+        return stream.pubseekoff(offset, way == SEEK_SET ? std::ios_base::beg : (way == SEEK_END ? std::ios_base::end : std::ios_base::cur));
+    }
+
+    IReadStream* clone() override
+    {
+        return new StreambufReadStream(stream);
```

Comment:
Looks useless because it could not (and should not) create independent objects (at least seek pointer is shared).

See below comment about lifetime.

Reply:
Removed `clone` method completely in last commit. As I understand, `struct CvCapture_FFMPEG` cannot hold a copy of `Ptr<IReadStream>` because there is no default constructor to allocate it: https://github.com/opencv/opencv/blob/b42075f3e2c7eca4a12e55b32cfc95fa90982a21/modules/videoio/src/cap_ffmpeg_impl.hpp#L3296

So I used `CvCapture_FFMPEG_proxy` to increase a counter